src/client/ds3/ds3_internal.h

```diff
@@ -58,6 +58,8 @@ struct ds3_bucket {
 struct ds3_obj {
 	/** DFS object handle */
 	dfs_obj_t *dfs_obj;
+	d_sg_list_t	ds3_sgl;
+	d_iov_t		ds3_iov;
```

Comment:
I assume that we could have concurrent read operations against the same S3 object. In this case, they will all use the same sgl and iov and causes corruption, no?

Reply:
that's reasonable given the ds3 API accepts an ev. although the sgl you are passing would replace the `void *buf` since the sgl is the memory buffer descriptor. If we do that, then should also update the ds3 write API to match.

alternatively one could allocate a per IO struct and free it on completion by registering a completion cb on the event. No strong preference on my side, but slight to the former.

Reply:
It sounds like libds3 users won't need to manage scatter-gather list, so it sounds to me that we should hide the complexity of managing the sgl from the user. I would personally then be in favor of keeping the API simple with just a buf and manage the sgl internally (possibly as in a completion callback as suggested by Mohamad).

Reply:
Presently, for every S3 PUT/GET operation a new `ds3_obj` is allocated, thus I think it is safe to embed iov and sgl in `ds3_obj`.

Reply:
Thanks. I still think that we should use a callback. @chucktuffli, is that something you could please look into? I think that you can use the dfs read path as an example: https://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c#L4646
Thanks in advance.

Reply:
The last push removed the SGL/IOV from the object and follows a pattern similar to `dfs_read` in that it a) uses stack variables for the synchronous case (i.e., event is NULL) and b) it allocates the SGL/IOV for the asynchronous case and frees the memory in a callback registered via `daos_event_register_comp_cb()`

Reply:
Thanks. Asking for folks to review it (otherwise it does not appear in review list).